### PR TITLE
`set_signature` API to set the signature for model artifacts

### PR DESCRIPTION
**Pull request by jerrylian-db**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

This PR introduces a `set_signature` API that makes it possible to set signatures for model artifacts. These model artifacts can be locally saved models, logged run model artifacts, or artifacts in any of the supported artifact stores. The function respects the immutability of model registry artifacts by disallowing model artifacts specified with a `model:/` scheme URI.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

I tested the `set_signature` API in a Databricks notebook. Here are the steps I took:
1. Installed my feature branch using `%pip install git+https://github.com/mlflow/mlflow@add_signature`
2. Logged a model with an MLflow signature. As shown below:
Screenshot 2023-05-19 at 10 40 23 AM
3. Set the signature of the model via the command `set_signature("runs:/{}/{}".format(run.info.run_id, "sklearn-model"), signature)`. Checked the run artifacts UI again and saw this:
Screenshot 2023-05-19 at 10 41 34 AM

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

Screenshot 2023-05-19 at 10 22 06 AM

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Introduces a `mlflow.models.set_signature()` API to set signatures for model artifacts.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/models/signature.py [hunk 1]
```diff
@@ -238,3 +245,76 @@ def _infer_signature_from_type_hints(func, input_arg_index, input_example=None):
...
+
+        # set the signature for the logged model
+        set_signature(model_uri, signature)
+    """
+    assert isinstance(
+        signature, ModelSignature
+    ), "The signature argument must be a ModelSignature object"
+    assert not ModelsArtifactRepository.is_models_uri(
+        model_uri
+    ), "Model URIs with the `models:/` scheme are not supported."
```
Comment (dbczumar) [inline mlflow/models/signature.py:304]: Can we turn these into `MlflowException` with error code `INVALID_PARAMETER_VALUE`?

Comment (dbczumar) [inline mlflow/models/signature.py:304]: Can we update the error message to recommend that users create a new model version? It's always nice when I see an error message that tells me what to do next :D

File: mlflow/models/signature.py [hunk 2]
```diff
@@ -238,3 +245,76 @@ def _infer_signature_from_type_hints(func, input_arg_index, input_example=None):
...
+    signature: ModelSignature,
+):
+    """
+    Sets the model signature for specified model artifacts.
+
+    The process involves downloading the MLmodel file in the model artifacts (if it's non-local),
+    updating its model signature, and then overwriting the existing MLmodel file. Should the
+    artifact repository associated with the model artifacts disallow overwriting, this function will
+    fail. Furthermore, as model registry artifacts are read-only, model artifacts located in the
+    model registry and represented by ``models:/`` URI schemes are not compatible with this API.
```
Comment (dbczumar) [inline mlflow/models/signature.py:261]: Can we recommend that users create a new model version in order to update the signature? Might be nice to include a link to some docs about how to do that (we may have to add those docs as part of this PR too, or feel free to inline as a Python code example)

File: tests/tracking/test_artifact_utils.py [hunk 3]
```diff
@@ -119,3 +120,22 @@ def test_upload_artifacts_to_databricks_no_run_id():
...
+
+def test_upload_artifacts_to_uri(tmpdir):
+    artifact_file_name = "artifact.txt"
+    artifact_text = "Sample artifact text"
+    local_artifact_path = tmpdir.join(artifact_file_name).strpath
+    with open(local_artifact_path, "w") as out:
+        out.write(artifact_text)
+
+    with mlflow.start_run() as run:
+        mlflow.log_metric("coolness", 1)
```
Comment (dbczumar) [inline tests/tracking/test_artifact_utils.py:133]: Coolness, indeed :D

**Review by dbczumar — APPROVED** [review]
LGTM with nitty docs comments. Impl looks wonderful! Thanks @jerrylian-db !

File: mlflow/models/signature.py (hunk 2 above)
Comment (jerrylian-db) [reply mlflow/models/signature.py:261]: I'll add a new doc section in the Model Signatures doc in a follow up PR.

File: mlflow/models/signature.py (hunk 1 above)
Comment (jerrylian-db) [reply mlflow/models/signature.py:304]: In a follow-up PR, I'll link to documentation that describes how to do this.

**Pull request merged**